src/types/nullable.rs

```diff
@@ -90,46 +94,103 @@ fn snapshot_take<'ctx, 'this>(
+    let block_finish = helper.append_block(Block::new(&[(pointer(context, 0), location)]));
 
-                block.append_operation(scf::r#yield(&[null_ptr], location));
-                region
-            },
-            {
-                let region = Region::new();
-                let block = region.append_block(Block::new(&[]));
+    entry.append_operation(cf::cond_br(
+        context,
+        is_null,
+        block_finish,
+        block_not_null,
+        &[null_ptr],
+        &[],
+        location,
+    ));
+
+    {
+        let value_len =
+            block_not_null.const_int(context, location, inner_layout.pad_to_align().size(), 64)?;
+
+        let dst_ptr = block_not_null.append_op_result(ReallocBindingsMeta::realloc(
+            context, null_ptr, value_len, location,
+        ))?;
 
-                let alloc_len = block.const_int(context, location, elem_layout.size(), 64)?;
+        match inner_snapshot_take {
+            Some(inner_snapshot_take) => {
+                let value = block_not_null.load(context, location, src_value, inner_ty)?;
 
-                let cloned_ptr = block.append_op_result(ReallocBindingsMeta::realloc(
-                    context, null_ptr, alloc_len, location,
-                ))?;
+                let (next_block, value) = inner_snapshot_take(
+                    context,
+                    registry,
+                    block_not_null,
+                    location,
+                    helper,
+                    metadata,
+                    value,
+                )?;
+                block_not_null = next_block;
 
-                block.append_operation(
+                block_not_null.store(context, location, dst_ptr, value)?;
+            }
+            None => {
+                block_not_null.append_operation(
                     ods::llvm::intr_memcpy(
                         context,
-                        cloned_ptr,
+                        dst_ptr,
                         src_value,
-                        alloc_len,
+                        value_len,
                         IntegerAttribute::new(IntegerType::new(context, 1).into(), 0),
                         location,
                     )
                     .into(),
                 );
+            }
+        }
+        block_not_null.append_operation(cf::br(block_finish, &[dst_ptr], location));
+    }
 
-                block.append_operation(scf::r#yield(&[cloned_ptr], location));
-                region
-            },
-            location,
-        ))
-        .result(0)?
-        .into();
+    let value = block_finish.argument(0)?.into();
+
+    Ok((block_finish, value))
+}
+
+#[cfg(test)]
+mod test {
+    use crate::{
+        utils::test::{jit_enum, jit_struct, load_cairo, run_program},
+        values::JitValue,
+    };
+    use pretty_assertions_sorted::assert_eq;
+
+    #[test]
+    fn test_nullable_deep_clone() {
+        let program = load_cairo! {
+            use core::array::ArrayTrait;
```

Comment:
one question: why can we add this test as a normal cairo program?

Reply:
We're testing if executing this program causes a segfault or not. If it causes a segfault it'll crash the test suite and therefore make the test "fail".

The test works by forcing the invocation of the fix (by taking a snapshot of the nullable), dropping the original value and returning the snapshot. Since its value is an array, which doesn't implement Copy, it'll need to be cloned because we're destroying the original therefore freeing the array. If the value was copied and not cloned then it'd cause a double free or segfault crash.